src/auth/webhook-secret.guard.ts

```diff
@@ -0,0 +1,32 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+import * as crypto from 'crypto';
+import { Request } from 'express';
+import { Env } from 'src/env';
+
+@Injectable()
+export class WebhookSecretGuard implements CanActivate {
+  canActivate(context: ExecutionContext): boolean {
+    const request: Request = context.switchToHttp().getRequest();
+    const signature = request.headers['x-hub-signature-256'] as string;
+    const secret = Env.SECRET.WEBHOOK;
+    const payload = JSON.stringify(request.body);
+
+    if (!signature) {
+      throw new UnauthorizedException('Missing signature');
+    }
+
+    const hmac = crypto.createHmac('sha256', secret);
+    const digest = `sha256=${hmac.update(payload).digest('hex')}`;
+
+    if (!crypto.timingSafeEqual(Buffer.from(digest), Buffer.from(signature))) {
+      throw new UnauthorizedException('Invalid signature');
+    }
+
+    return true;
+  }
```

Comment:
PullProofAI

This `WebhookSecretGuard` class looks well-implemented for verifying webhook signatures. However, there are a few improvements that can be suggested:

1. Consider validating the presence of the `x-hub-signature-256` header in a separate method for better separation of concerns.
2. Ensure error handling is robust and covers edge cases like invalid payloads or secrets.
3. Add comments to explain the purpose of critical sections of code for better maintainability.

```diff
+ // Consider extracting header validation into a separate method
+ // Improve error handling for edge cases
+ // Add comments for clarity and maintainability
```



<!-- This is an auto-generated comment by PullProofAI -->